src/apify/_actor.py

```diff
@@ -839,12 +841,29 @@ async def reboot(
             self.log.error('Actor.reboot() is only supported when running on the Apify platform.')
             return
 
+        if self._is_rebooting:
+            self.log.debug('Actor is already rebooting, skipping the additional reboot call.')
+            return
+
+        self._is_rebooting = True
+
         if not custom_after_sleep:
             custom_after_sleep = self._configuration.metamorph_after_sleep
 
-        self._event_manager.emit(event=Event.PERSIST_STATE, event_data=EventPersistStateData(is_migrating=True))
+        # Call all the listeners for the PERSIST_STATE and MIGRATING events, and wait for them to finish.
+        # We can't just emit the events and wait for all listeners to finish,
+        # because this method might be called from an event listener itself, and we would deadlock.
+        persist_state_listeners = chain.from_iterable(
+            (self._event_manager._listeners_to_wrappers[Event.PERSIST_STATE] or {}).values()  # noqa: SLF001
+        )
+        migrating_listeners = chain.from_iterable(
+            (self._event_manager._listeners_to_wrappers[Event.MIGRATING] or {}).values()  # noqa: SLF001
+        )
```

Comment:
I might be missing something, but the `chain.from_iterable` doesn't seem to have a purpose here.

Reply:
Huh, thanks. SDK already depends on Crawlee and that depends on `more_itertools` which contain `flatten` - could you add that dependency to SDK as well and use that instead? I know it serves no practical purpose other than readability.

Reply:
Sure, will do.